core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/SigningStage.java

```diff
@@ -74,28 +74,22 @@ private <T extends Identity> SdkHttpFullRequest sraSignRequest(SdkHttpFullReques
 
         ExecutionAttributes executionAttributes = context.executionAttributes();
         SelectedAuthScheme<T> selectedAuthScheme =
-            executionAttributes.getAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME);
+            (SelectedAuthScheme<T>) executionAttributes.getAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME);
```

Comment:
If you pass the selected auth scheme as a separate parameter to the method, no casting is needed because T will be the selected auth scheme's type parameter:

```java
return sraSignRequest(request, context, context.executionContext().getAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME));
```
```java
 private <T extends Identity> SdkHttpFullRequest sraSignRequest(SdkHttpFullRequest request, RequestExecutionContext context, SelectedAuthScheme<T> selectedAuthScheme);
```

You could also do that within the sraSignRequest method (if you want to keep the sraSignRequest signature simple) by delegating to yet another method (doSraSignRequest?). Essentially, you're trying to assign a name (T) to the ? in SelectedAuthScheme so that you can use it to refer to the type the select auth scheme contains. You can do that with any method call.

Reply:
Yeah, the issue was making that T work for the Identity, which we discuss below.